AnnService/inc/Helper/DiskIO.h

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef _SPTAG_HELPER_DISKIO_H_
+#define _SPTAG_HELPER_DISKIO_H_
+
+#include <functional>
+#include <fstream>
+#include <string.h>
+
+namespace SPTAG
+{
+    namespace Helper
+    {
+        enum class DiskIOScenario
+        {
+            DIS_BulkRead = 0,
+            DIS_UserRead,
+            DIS_HighPriorityUserRead,
+            DIS_Count
+        };
+
+        struct AsyncReadRequest
+        {
+            std::uint64_t m_offset;
+            std::uint64_t m_readSize;
+            char* m_buffer;
+            std::function<void(bool)> m_callback;
+            
+            // Carry items like counter for callback to process.
+            void* m_payload;
+            bool m_success;
+
+            AsyncReadRequest() : m_offset(0), m_readSize(0), m_buffer(nullptr), m_payload(nullptr), m_success(false) {}
+        };
+
+        class DiskPriorityIO
+        {
+        public:
+            DiskPriorityIO(DiskIOScenario scenario = DiskIOScenario::DIS_UserRead) {}
+
+            virtual ~DiskPriorityIO() {}
+
+            virtual bool Initialize(const char* filePath, int openMode,
+                // Max read/write buffer size.
+                std::uint64_t maxIOSize = (1 << 20),
+                std::uint32_t maxReadRetries = 2,
+                std::uint32_t maxWriteRetries = 2,
+                std::uint16_t threadPoolSize = 4) = 0;
+
+            virtual std::uint64_t ReadBinary(std::uint64_t readSize, char* buffer, std::uint64_t offset = UINT64_MAX) = 0;
+
+            virtual std::uint64_t WriteBinary(std::uint64_t writeSize, const char* buffer, std::uint64_t offset = UINT64_MAX) = 0;
+
+            virtual std::uint64_t ReadString(std::uint64_t& readSize, std::unique_ptr<char[]>& buffer, char delim = '\n', std::uint64_t offset = UINT64_MAX) = 0;
+
+            virtual std::uint64_t WriteString(const char* buffer, std::uint64_t offset = UINT64_MAX) = 0;
+
+            virtual bool ReadFileAsync(AsyncReadRequest& readRequest) = 0;
+
+            virtual std::uint64_t TellP() = 0;
```

Comment:
since our interface is like :
virtual unsigned __int32 WriteFile(unsigned __int64 offset, unsigned __int32 writeSize, __int8* buffer) = 0;

every read write will return actual file pointer relative position, you can get position by adding offset

Reply:
We return the actual read or write bytes.